bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/directentrylogger/DirectWriter.java

```diff
@@ -0,0 +1,322 @@
+                                          .kv("filename", filename)
+                                          .kv("writeSize", bytesToWrite)
+                                          .kv("errno", ne.getErrno())
+                                          .kv("offset", offset).toString());
+                } finally {
+                    bufferPool.release(tmpBuffer);
+                }
+                return null;
+            });
+        addOutstandingWrite(f);
+    }
+
+    @Override
+    public int writeDelimited(ByteBuf buf) throws IOException {
+        synchronized (bufferLock) {
+            if (!nativeBuffer.hasSpace(serializedSize(buf))) {
+                flushBuffer();
+            }
+
+            int readable = buf.readableBytes();
+            long bufferPosition = position() + Integer.BYTES;
+            if (bufferPosition > Integer.MAX_VALUE) {
+                throw new IOException(exMsg("Cannot write past max int")
+                                      .kv("filename", filename)
+                                      .kv("writeSize", readable)
+                                      .kv("position", bufferPosition)
+                                      .toString());
+            }
+            nativeBuffer.writeInt(readable);
+            nativeBuffer.writeByteBuf(buf);
+            return (int) bufferPosition;
+        }
+    }
+
+    @Override
+    public void position(long offset) throws IOException {
+        synchronized (bufferLock) {
+            if (nativeBuffer != null && nativeBuffer.position() > 0) {
+                flushBuffer();
+            }
+            if ((offset % Buffer.ALIGNMENT) != 0) {
+                throw new IOException(exMsg("offset must be multiple of alignment")
+                                      .kv("offset", offset)
+                                      .kv("alignment", Buffer.ALIGNMENT)
+                                      .toString());
+            }
+            this.offset = offset;
+        }
+    }
+
+    @Override
+    public long position() {
+        synchronized (bufferLock) {
+            return this.offset + (nativeBuffer != null ? nativeBuffer.position() : 0);
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        flushBuffer();
+
+        waitForOutstandingWrites();
+
+        try {
+            int ret = nativeIO.fsync(fd);
+            checkState(ret == 0, "Fsync should throw exception on non-zero return (%d)", ret);
+        } catch (NativeIOException ne) {
+            throw new IOException(exMsg(ne.getMessage())
+                                  .kv("file", filename)
+                                  .kv("errno", ne.getErrno()).toString());
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        synchronized (bufferLock) {
+            if (nativeBuffer != null && nativeBuffer.position() > 0) {
+                flushBuffer();
+            }
+        }
+
```

Comment:
I think you should do `waitForOutstandingWrites();` here or call `flush()` instead of `flushBuffer()` to confirm that the last outstanding write (submitted e.g. by `flushBuffer()`) actually succeeded.